### PR TITLE
feat(sdk-core): Phase 5 of GG18 signing

### PR DESCRIPTION
**Pull request by zhongxishen**
HSM-74

 - Hash commitment
 - Schnorr NIZK proof
 - NIZK proof of V value calculated in step 5A of GG18 signing
 - New functions of Ecdsa class to support phase 5
 - Tests that demonstrate signing with phase 5

### CODE REVIEW

**Review by zahin-mohammad — COMMENTED** [review]
Would like to get an HSM team review on this first before taking a look at it 🙏🏾

File: modules/sdk-lib-mpc/test/unit/schnorrProof.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import 'should';
```
Comment (OttoAllmendinger) [inline modules/sdk-lib-mpc/test/unit/schnorrProof.ts:1]: We need to deprecate `should`. It cannot handle `BigInt`: `({a: 1n}).should.eql({a: 2n})` will pass.

**Review by OttoAllmendinger — COMMENTED** [review]
not enough context to review

**Review by islamaminBitGo — DISMISSED** [review]
Crypto functions LGTM

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts [hunk 2]
```diff
@@ -1305,9 +1313,9 @@ export default class Ecdsa {
    * @param {DShare} dShare delta share received from the other participant
    * @param {Hash} hash hashing algorithm implementing Node`s standard crypto hash interface
    * @param {boolean} shouldHash if true, we hash the provided buffer before signing
-   * @returns {SShare}
+   * @returns {VAShare}
    */
-  sign(M: Buffer, oShare: OShare, dShare: DShare, hash?: Hash, shouldHash = true): SShare {
+  sign(M: Buffer, oShare: OShare, dShare: DShare, hash?: Hash, shouldHash = true): VAShare {
```
Comment (zahin-mohammad) [inline modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:1318]: Will this break the existing signing flow for clients or anyone using the latest sdk?

**Review by zahin-mohammad — COMMENTED** [review]
all my suggestions are non-blocking. 

`sdk-lib-mpc` changes lgtm, but have some concerns regarding `sdk-core`

File: modules/sdk-lib-mpc/src/tss/ecdsa/types.ts [hunk 3]
```diff
@@ -226,3 +226,9 @@ export interface RangeProofWithCheck {
   t2: bigint;
   u: bigint;
 }
+
+export interface ZkVProof {
+  Alpha: bigint;
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/src/tss/ecdsa/types.ts:231]: This is intentional, capitalization means it's point value on the curve.

File: modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts (hunk 2 above)
Comment (zhongxishen) [reply modules/sdk-core/src/account-lib/mpc/tss/ecdsa/ecdsa.ts:1318]: As far as I know, it shouldn't as `VAShare` is subtype of `SShare`.  In other words, everything from `SShare` is there and remains the same.

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import 'should';
+import { randomBytes } from 'crypto';
+import { Secp256k1Curve } from '../../../../src/curves';
+import { createZkVProof, verifyZkVProof } from '../../../../src/tss/ecdsa/zkVProof';
+
+describe('zkV proof', function () {
+  const curve = new Secp256k1Curve();
+
+  it('create and verify proof', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    verifyZkVProof(V, createZkVProof(V, s, l, R, curve), R, curve).should.be.true();
+
+    const sessionId = randomBytes(32);
+    const zkVProof = createZkVProof(V, s, l, R, curve, sessionId);
+
+    verifyZkVProof(V, zkVProof, R, curve, sessionId).should.be.true();
+  });
+
+  it('proof not verify with partial V', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    const R = curve.basePointMult(curve.scalarRandom());
+    const V = curve.pointAdd(curve.pointMultiply(R, s), curve.basePointMult(l));
+    const partialV = curve.pointMultiply(R, s); // partial V is missing the lG term
+
+    const zkVProof = createZkVProof(V, s, l, R, curve);
+
+    verifyZkVProof(partialV, zkVProof, R, curve).should.be.false();
+  });
+
+  it('proof created with bad s not verify', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    const R = curve.basePointMult(curve.scalarRandom());
+    const V = curve.pointAdd(curve.pointMultiply(R, s), curve.basePointMult(l));
+
+    const s2 = curve.scalarRandom();
+    const zkVProof = createZkVProof(V, s2, l, R, curve);
+
+    verifyZkVProof(V, zkVProof, R, curve).should.be.false();
+  });
+
+  it('proof not verify with wrong contextual info', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/tss/ecdsa/zkVProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/schnorrProof.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import { Secp256k1Curve } from '../../src/curves';
+import { createSchnorrProof, verifySchnorrProof } from '../../src/schnorrProof';
+
+describe('schnorr proof', function () {
+  const curve = new Secp256k1Curve();
+
+  it('create and verify proof', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/schnorrProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/schnorrProof.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    verifySchnorrProof(A, createSchnorrProof(A, a, curve), curve).should.be.true();
+
+    const m = 'sign this';
+    const schnorrProof = createSchnorrProof(A, a, curve, Buffer.from(m));
+
+    verifySchnorrProof(A, schnorrProof, curve, Buffer.from(m)).should.be.true();
+  });
+
+  it('incorrect proof', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/schnorrProof.ts]: done

File: modules/sdk-lib-mpc/test/unit/hashCommitment.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    const comDecom = createCommitment(Buffer.concat([buf1, buf2]));
+
+    verifyCommitment(comDecom.commitment, comDecom.decommitment).should.be.true();
+
+    const comDecom2 = createCommitment(Buffer.concat([buf1, buf2]), Buffer.concat([randomBytes(32), randomBytes(32)]));
+
+    verifyCommitment(comDecom2.commitment, comDecom2.decommitment).should.be.true();
+  });
+
+  it('invalid parameters', function () {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/test/unit/hashCommitment.ts]: done

File: modules/sdk-lib-mpc/src/tss/ecdsa/types.ts [hunk 11]
```diff
@@ -226,3 +226,9 @@ export interface RangeProofWithCheck {
   t2: bigint;
   u: bigint;
 }
+
+export interface ZkVProof {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/src/tss/ecdsa/types.ts:230]: Sure, maybe in a follow up PR?

File: modules/sdk-lib-mpc/src/tss/ecdsa/zkVProof.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  const u = curve.scalarAdd(b, curve.scalarMult(c, l));
+
+  return {
+    Alpha: Alpha,
+    t: t,
+    u: u,
+  };
+}
+
+function nonInteractiveChallenge(V: bigint, R: bigint, Alpha: bigint, curve: Curve, additionalCtx: Buffer): bigint {
```
[the remark this answers is not in the document]
Comment (zhongxishen) [reply modules/sdk-lib-mpc/src/tss/ecdsa/zkVProof.ts:55]: done

**Comment by zhongxishen** [comment]
rebased after one test failed in `sdk-coin-trx`

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**